GitDiffMargin/Git/GitCommands.cs

```diff
@@ -282,5 +325,90 @@ static bool HasPreamble(string file, Encoding encoding)
 
             return true;
         }
+
+        private string AdjustPath(string fullPath)
+        {
+            // Right now the only extension is for *.csproj and *.vbproj, which are opened from a temporary location
+            switch (Path.GetExtension(fullPath).ToLowerInvariant())
+            {
+            case ".csproj":
+            case ".vbproj":
+                break;
+
+            default:
+                return fullPath;
+            }
+
+            // CPS will open the file in %TEMP%\{random name}\{ProjectFileName}
+            string directoryName = Path.GetDirectoryName(fullPath);
+            if (string.IsNullOrEmpty(directoryName))
+                return fullPath;
+
+            directoryName = Path.GetDirectoryName(directoryName);
+            if (!Path.GetTempPath().Equals(directoryName + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase))
+                return fullPath;
+
+            IVsSolution solution = _serviceProvider.GetService(typeof(SVsSolution)) as IVsSolution;
+            if (solution == null)
+                return fullPath;
+
+            if (!ErrorHandler.Succeeded(solution.GetProjectEnum((uint)__VSENUMPROJFLAGS.EPF_LOADEDINSOLUTION, Guid.Empty, out IEnumHierarchies ppenum))
+                || ppenum == null)
+            {
+                return fullPath;
+            }
+
+            List<string> projectFiles = new List<string>();
+            IVsHierarchy[] hierarchies = new IVsHierarchy[1];

```

Comment:
I doubt there are guaranteed to be 2 hierarchies, especially  if it's a solution with no projects.